docs/resources/server/recipes/index.md

```diff
@@ -226,11 +211,296 @@ public class RightClickBlockRecipe implements Recipe<RightClickBlockInput> {
+    )
+}
+```
+
+### Placement Info
+
+A `PlacementInfo` is meant to define the crafting requirements used by the recipe consumer to determine what recipes can be placed in workbench-type. `PlacementInfo` are only meant for item ingredients, so if other types of ingredients are desired (e.g., fluid, block), the surrounding logic will need to be implemented from scratch. In these cases, the recipe can be labelled as not placeable, and say as such via `PlacementInfo#NOT_PLACEABLE`. However, if there is at least one item-like object in your recipe, you should create a `PlacementInfo`.
+
+A `PlacementInfo` can be created via `create`, which takes in one or a list of ingredient, or `createFromOptionals`, which takes in a list of optional ingredients. If your recipe contains some representation of empty slots, then `createFromOptionals` should be used, providing an empty optional for an empty slot:
+
+```java
+public class RightClickBlockRecipe implements Recipe<RightClickBlockInput> {
+    // other stuff here
+    private PlacementInfo info;
+
+    @Override
+    public PlacementInfo placementInfo() {
+        // This delegate is in case the ingredient is not fully populated at this point in time
+        // Tags and recipes are loaded at the same time, which is why this might be the case.
+        if (this.info == null) {
+            // Use optional ingredient as the block state may have an item representation
+            List<Optional<Ingredient>> ingredients = new ArrayList<>();
+            Item stateItem = this.inputState.getBlock().asItem();
+            ingredients.add(stateItem != Items.AIR ? Optional.of(Ingredient.of(stateItem)): Optional.empty());
+            ingredients.add(Optional.of(this.inputItem));
+
+            // Create placement info
+            this.info = PlacementInfo.createFromOptionals(ingredients);
+        }
+
+        return this.info;
+    }
+}
+```
+
+### Slot Displays
+
+`SlotDisplay`s represent the information on what should render in what slot when viewed by a recipe consumer, like a recipe book. A `SlotDisplay` has two methods. First there's `resolve`, which takes in the `ContextMap` containing the available registries and fuel values (as shown in `SlotDisplayContext`); and the current `DisplayContentsFactory`, which accepts the contents to display for this slot; and returns the transformed list of contents into the output to be accepted. Then there's `type`, which holds the [`MapCodec`][codec] and [`StreamCodec`][streamcodec] used to encode/decode the display.
+
+`SlotDisplay`s are typically implemented on the [`Ingredient` via `#display`, or `ICustomIngredient#display` for modded ingredients][ingredients]; however, in some cases, the input may not be an ingredient, meaning a `SlotDisplay` will need to use one available, or have a new one created.
+
+These are the available slot displays provided by Vanilla and NeoForge:
+
+- `SlotDisplay.Empty`: A slot that represents nothing.
```

Comment:
This is displayed weirdly, check the preview. I think it's because of the blank lines between two list elements? This also affects the `DisplayContentsFactory` list below.